tracr/craft/bases.py

```diff
@@ -135,48 +163,96 @@ def project(
     if isinstance(basis, VectorSpaceWithBasis):
       basis = basis.basis
     components = []
+    # use a set for lookups, greatly improves performance
+    basis_set = self.get_basis_set()
+    direction_to_index = self.get_direction_to_index()
     for direction in basis:
-      if direction in self.basis_directions:
+      if direction in basis_set:
         components.append(
-            self.magnitudes[..., self.basis_directions.index(direction)])
+            self.magnitudes[..., direction_to_index[direction]])
       else:
         components.append(np.zeros_like(self.magnitudes[..., 0]))
-    return VectorInBasis(list(basis), np.stack(components, axis=-1))
-
-
+    
+    if isinstance(basis, VectorSpaceWithBasis): # we can save compute and construct the vector in basis from the vector space
+      return basis.to_vector_in_basis(np.stack(components, axis=-1))
+    else: # otherwise construct one from scratch, we may have to build the set and index dict if requested
+      return VectorInBasis(list(basis), np.stack(components, axis=-1), basis_is_sorted=True)
+  
+  def add_directions(self, vector: "VectorInBasis") -> "VectorInBasis":
+    """
+    Equivalent to self += vector.project(self.basis)
+    This is to speed up accumulation into an output vector in the combine_in_parallel in vectorspace_fns
+    """
+    directions_indices = self.get_direction_to_index()
+    new_magnitudes = np.array(self.magnitudes)
+    for idx, basis in enumerate(vector.basis_directions):
+      new_magnitudes[..., directions_indices[basis]] += vector.magnitudes[..., idx]
+    return self.update_magnitudes(new_magnitudes)
+  
+  def update_magnitudes(self, new_magnitudes):
+    vect = VectorInBasis(list(self.basis_directions), new_magnitudes, basis_is_sorted=True)
+    vect._basis_set = self._basis_set
+    vect._direction_to_index = self._direction_to_index
+    return vect
+  
 @dataclasses.dataclass
 class VectorSpaceWithBasis:
-  """A vector subspace in a given basis."""
+  """A vector subspace in a given basis.
+  
+  _basis_set and _direction_to_index should be left None
+  """
   basis: Sequence[BasisDirection]
+  _basis_set: Optional[set] = None
+  _direction_to_index: Optional[dict] = None
 
   def __post_init__(self):
     """Keep basis directions sorted."""
     self.basis = sorted(self.basis)
-
+    
+  def to_vector_in_basis(self, magnitudes):
```

Comment:
This sounds like it's converting a vector space into a vector which doesn't make sense. Maybe better "make_vector" or similar

Reply:
Also here and elsewhere please add type annotations and run pylint and pytype